design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,499 @@
+# Component mutable and versioning mechanism
+
+* Owner: Lei Zhang (@resouer), Jianbo Sun (@wonderflow)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Terminology
+
+* Application: A holistic term for talking about all pieces that are rendered from an `ApplicationConfiguration`.
+* ComponentSchematic: A term in OAM spec v1alpha1, which is deprecated and replaced by `Component` in v1alpha2.
+
+## Background
+
+OAM v1alpha1 made a requirement that developer need to define another ComponentSchematic with
+different name for an update (i.e. ComponentSchematic is immutable). We've got lots of feedback from
+real practices in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
+that it's not a good experience. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356),
+this proposal is going to make it into implementation.
+
+Since component is becoming mutable, a very common question is how will the component versioning mechanism work.
+
+One of the requirement is the versioning mechanism should allow users to specify a fixed version of component
+in ApplicationConfiguration, so that the component can be upgraded as before.
+
+On the other hand, the versioning mechanism should also be used by traits. For example, a trait can specify
+different revision and control deployment process. 
+
+
+## Goals
+
+1. Component is mutable. Developers who modify the Component object will directly trigger immediate
+upgrade of any `ApplicationConfiguration` that does not specify `Component` revision. 
+2. Component revision can be specified in ApplicationConfiguration. Operator can specify a fixed revision
+of Component in ApplicationConfiguration. Component mutation won't affect Application directly in this case, 
+and can be upgraded in a more convenient way by working with traits.
+3. The versioning mechanism should be compatible with current existing K8s workloads, eg, `K8s Deployment`, `OpenKruise`.
+  OAM should support two common ways used in community for versioning mechanism. 
+  - Workload itself can handle version. This means we can just update `Deployment` and let the workload do rollout or rollback without traits.
+  - Corporate with traits. OAM runtime should create different `Deployment` instances for different revision and let the traits do rollout or rollback.
+
+## Non-Goals
+
+This proposal will not give a specific rollout or rollback design, that can be done by traits.  
+
+## Proposal
+
+### Using ControllerRevision to record versions of Component
+
+Because a `Component` is mutable, we introduce an independent Revision object called [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision)
```

Comment:
```suggestion
Because a `Component` is mutable, we borrow an independent Revision object called [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) from the existing Kubernetes resources.
```